src/Domain/Tags.hpp

```diff
@@ -46,32 +50,254 @@ struct ElementMap : db::DataBoxTag {
+struct InterfaceBase;
+
+/// \ingroup DataBoxTags
+/// \ingroup ComputationalDomain
+/// \brief Prefix for an object on interfaces.
+///
+/// The contained object will be a map from ::Direction to the item
+/// type of `Tag`, with the set of directions being those produced by
+/// `DirectionsTag`.  If `Tag` is a compute item, its function will be
+/// applied separately to the data on each interface.  If some of the
+/// compute item's inputs should be taken from the volume even when
+/// applied on a slice, it may indicate them using a `volume_tags`
+/// typelist in the tag struct.
+///
+/// \tparam DirectionsTag the item of directions
+/// \tparam Tag the tag labeling the item
+template <typename DirectionsTag, typename Tag>
+struct Interface : InterfaceBase<DirectionsTag, Tag, Tag> {};
+
+namespace Interface_detail {
+// Check if a typelist contains a value.  Should be put somewhere more general.
+template <typename Sequence, typename Item>
+using contains = tmpl::found<Sequence, std::is_same<tmpl::_1, tmpl::pin<Item>>>;
+template <typename Sequence, typename Item>
+constexpr const bool contains_v = contains<Sequence, Item>::value;
+
+// Pull volume_tags from BaseComputeItem, defaulting to an empty list.
+template <typename BaseComputeItem, typename = cpp17::void_t<>>
+struct volume_tags {
+  using type = tmpl::list<>;
+};
+
+template <typename BaseComputeItem>
+struct volume_tags<BaseComputeItem,
+                   cpp17::void_t<typename BaseComputeItem::volume_tags>> {
+  using type = typename BaseComputeItem::volume_tags;
+};
+
+// Add an Interface wrapper to a tag if appropriate
+template <typename DirectionsTag, typename Tag, typename VolumeTags>
+using interface_compute_item_argument_tag =
+    tmpl::conditional_t<contains_v<VolumeTags, Tag>, Tag,
+                        Interface<DirectionsTag, Tag>>;
+
+// Compute the argument tags for the interface version of a compute item.
+template <typename DirectionsTag, typename BaseComputeItem>
+using interface_compute_item_argument_tags = tmpl::transform<
+    typename BaseComputeItem::argument_tags,
+    tmpl::bind<interface_compute_item_argument_tag, DirectionsTag, tmpl::_1,
+               tmpl::pin<typename volume_tags<BaseComputeItem>::type>>>;
+
+// Pull the direction's entry from interface arguments, passing volume
+// arguments through unchanged.
+template <bool IsVolumeTag>
+struct unmap_interface_args;
+
+template <>
+struct unmap_interface_args<true> {
+  template <size_t VolumeDim, typename T>
+  static constexpr decltype(auto) apply(
+      const ::Direction<VolumeDim>& /*direction*/, const T& arg) noexcept {
+    return arg;
+  }
+};
+
+template <>
+struct unmap_interface_args<false> {
+  template <size_t VolumeDim, typename T>
+  static constexpr decltype(auto) apply(const ::Direction<VolumeDim>& direction,
+                                        const T& arg) noexcept {
+    return arg.at(direction);
+  }
+};
+
+template <typename, typename, typename>
```

Comment:
Please name things in forward declarations. It gets really annoying trying to find the "correct" definition to figure out what the template parameters are supposed to mean.